refactor(jellyfin): abstract jellyfin hostname, updated ui to reflect it, better validation

fallenbagel opened the pull request:
> [!WARNING]
BREAKING CHANGE: Jellyfin settings now does not include a hostname. Instead it abstracted it to ip, port, useSsl, and urlBase. However, migration of old settings to new settings should work automatically.
#### Description
This PR refactors and abstracts jellyfin hostname into, jellyfin ip, jellyfin port, jellyfin useSsl, and jellyfin urlBase. This makes it more consistent with how plex settings are stored as well. In addition, this improves validation as validation can be applied seperately to them instead of as one whole regex doing the work to validate the url.
UI was updated to reflect this.

#### Screenshot (if UI-related)
![image](https://github.com/Fallenbagel/jellyseerr/assets/98979876/6c2d25df-d72a-41da-9acd-ce1b8e796596)

#### To-Dos

- [x] Successful build `yarn build`
- [x] Translation keys `yarn i18n:extract`
~- [ ] Database migration (if required)~
- [x] Abstract settings migration logic
- [x] Figure out why cypress tests are failing

#### Issues Fixed or Closed

- Fixes #XXXX

fallenbagel committed:
refactor(jellyfinsettings): abstract jellyfin hostname, updated ui to reflect it, better validation

This PR refactors and abstracts jellyfin hostname into, jellyfin ip, jellyfin port, jellyfin useSsl,
and jellyfin urlBase. This makes it more consistent with how plex settings are stored as well. In
addition, this improves validation as validation can be applied seperately to them instead of as one
whole regex doing the work to validate the url.
UI was updated to reflect this.

BREAKING CHANGE: Jellyfin settings now does not include a hostname. Instead it abstracted it to ip,
port, useSsl, and urlBase. However, migration of old settings to new settings should work
automatically.

gauthier-th commented on server/constants/error.ts:
```diff
@@ -2,4 +2,7 @@ export enum ApiErrorCode {
   NotAdmin = 'NOT_ADMIN',
+  SyncErrorGroupedFolders = 'SYNC_ERROR_GROUPED_FOLDERS',
+  SyncErrorNoLibraries = 'SYNC_ERROR_NO_LIBRARIES',
+  Generic = 'GENERIC',
```
This will cause merge conflict

Fallenbagel replied:
Oh yeah I'll resolve it

fallenbagel committed:
Merge remote-tracking branch 'origin/develop' into refactor-jellyfin-settings

fallenbagel committed:
refactor: remove console logs and use getHostname and ApiErrorCodes

fallenbagel committed:
fix: store req.body jellyfin settings temporarily and store only if valid

This should fix the issue where settings are saved even if the url
was invalid. Now the settings will only be saved if the url is
valid. Sort of like a test connection.

fallenbagel committed:
refactor: clean up commented out code

fallenbagel committed:
refactor(i18n): extract translation keys

gauthier-th reviewed (dismissed):
Works on my config. As it is a critical breaking change, could we have more people testing and reviewing this?

Fallenbagel commented:
> Works on my config. As it is a critical breaking change, could we have more people testing and reviewing this?

I will be creating a preview tag for this and try to get people to test this

Fallenbagel commented:
This pr should be available to test using the tag:
`preview-pr773`

fallenbagel committed:
fix(auth): auth failing with jellyfin login is disabled

Fallenbagel commented:
> In console log I get this
> 
> ```
> "Jellyfin login is disabled"
> ```
> 
> ```
> "cookie 'connect.sid' required"
> ```

A new preview tag should be available in about half an hour with a fix (when all the checks in this PR turns green). You'll just need to pull it.

sgtcoder commented:
Sounds good, let me know when it's updated.

fallenbagel committed:
Merge remote-tracking branch 'origin/develop' into refactor-jellyfin-settings

Gauvino reviewed (approved):
only migration problem but, other else looks good !

fallenbagel committed:
fix(settings): jellyfin migrations replacing the rest of the settings

fallenbagel committed:
fix(settings): jellyfin hostname should be carried out if hostname exists

fallenbagel committed:
fix(settings): merging the wrong settings source

sgtcoder commented:
I can check this out later today. I tried 3 days ago and it was having the same issue.

Fallenbagel commented:
> I can check this out later today. I tried 3 days ago and it was having the same issue.

There was a migration issue. I did not fix it 3 days ago. 
I just did.

fallenbagel committed:
refactor(settings): use migrator for dynamic settings migrations

fallenbagel committed:
refactor(settingsmigrator): settings migration handler and the migrations

fallenbagel committed:
test(cypress): fix cypress tests failing

cypress settings were lacking some of the jobs so when the startJobs() is called when the app
starts, it was failing to schedule the jobs where their cron timings were not specified in the
cypress settings. Therefore, this commit adds those jobs back. In addition, other setting options
were added to keep cypress settings consistent with a normal user.

fallenbagel committed:
chore(prettierignore): ignore cypress/config/settings.cypress.json as it does not need prettier

fallenbagel committed:
chore(prettier): ran formatter on cypress config to fix format check error

format check locally passes on this file. However, it fails during the github actions format check.
Therefore, json language features formatter was run instead of prettier to see if that fixes the
issue.

fallenbagel committed:
test(cypress): add only missing jobs to the cypress settings

fallenbagel committed:
Merge remote-tracking branch 'origin/develop' into refactor-jellyfin-settings

fallenbagel committed:
ci: attempt at trying to get formatter to pass on cypress config json file

fallenbagel committed:
Merge remote-tracking branch 'origin/develop' into refactor-jellyfin-settings

fallenbagel committed:
refactor: revert the changes brought to try and fix formatter

added back the rest of the cypress settings and removed cypress settings from .prettierignore

gauthier-th commented:
@sgtcoder could you test it one more time please?

gauthier-th commented on server/routes/settings/index.ts:
```diff
@@ -252,11 +255,59 @@ settingsRoutes.get('/jellyfin', (_req, res) => {
 });
 
-settingsRoutes.post('/jellyfin', (req, res) => {
+settingsRoutes.post('/jellyfin', async (req, res, next) => {
```
I don't really understand what you're doing here. What is the point of all this new code?

Fallenbagel replied:
This is relevant to how the new abstracted jellyfin hostname settings are saved. Previously when it was being saves, it won't do a check. It just saves. Even if it's wrong.

Now it will check if the url is reachable before saving. This is consistent with how plex settings are saved if you scroll up a little

gauthier-th replied:
Ok, thanks for the clarification. Make sense.

gauthier-th commented on server/routes/settings/index.ts:
```diff
@@ -252,11 +255,59 @@ settingsRoutes.get('/jellyfin', (_req, res) => {
+    const result = await jellyfinClient.getSystemInfo();
+
+    if (!result?.Id) {
+      throw new ApiError(result?.status, ApiErrorCode.InvalidUrl);
```
Why throw an ApiError to finally catch it just after?

gauthier-th replied:
Yes I think it's cleaner this way.

gauthier-th commented on server/routes/settings/index.ts:
```diff
@@ -252,11 +255,59 @@ settingsRoutes.get('/jellyfin', (_req, res) => {
+
+      return next({
+        status: 500,
+        message: 'Something went wrong',
```
Don't we have an ApiErrorCode for this?

gauthier-th commented on cypress/config/settings.cypress.json:
Is this linked to this PR?

gauthier-th replied:
Even all the `schedule` properties?

Fallenbagel replied:
> Even all the `schedule` properties?

Yes. Basically settings were missing all those and it made cypress tests fail only on this pr. Every other pr doesn't seem to care. Because i don't think the startJobs() was working until this pr when we abstracted how settings worked

fallenbagel committed:
Merge remote-tracking branch 'origin/develop' into refactor-jellyfin-settings

fallenbagel committed:
refactor(settings): better erorr logging when jellyfin connection test fails in settings page

Fallenbagel commented:
:tada: This PR is included in version 2.0.0 :tada:

The release is available on:
- `v2.0.0`
- [GitHub release](https://github.com/Fallenbagel/jellyseerr/releases/tag/v2.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
